### PR TITLE
bpo-26103 resolve Data descriptor contradiction

### PR DESCRIPTION
**Pull request by aaronchall**
https://bugs.python.org/issue26103

### CODE REVIEW

File: Lib/test/test_inspect.py [hunk 1]
```diff
@@ -1077,6 +1077,51 @@ class C(metaclass=M):
         attrs = [a[0] for a in inspect.getmembers(C)]
         self.assertNotIn('missing', attrs)
 
+class TestIsDataDescriptor(unittest.TestCase):
+
+    def test_custom_descriptors(self):
+        class NonDataDescriptor(object):
```
Comment (matrixise) [inline Lib/test/test_inspect.py]: Why do you specify the `object` class for the parent ? it is not required https://docs.python.org/3/reference/compound_stmts.html#class-definitions

Comment (aaronchall) [reply Lib/test/test_inspect.py]: This could be backported to 2, it is ostensibly fixing an incorrect implementation.

**Comment by aaronchall** [comment]
As it currently stands I am perfectly happy with this revision, but I am more than willing to consider change requests.

File: Lib/test/test_inspect.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/test/test_inspect.py]: Keep the Python 3 code clear. Add the `object` class in a backport to Python 2.

File: Lib/test/test_inspect.py [hunk 2]
```diff
@@ -1077,6 +1077,51 @@ class C(metaclass=M):
...
+        self.assertTrue(inspect.isdatadescriptor(type(Slotted.foo)),
+                        'a slot is a data descriptor')
+
+    def test_property(self):
+        self.assertTrue(inspect.isdatadescriptor(property),
+                        'property is a data descriptor')
+
+    def test_functions(self):
+        class Test(object):
+            def instance_method(self): ''
```
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: Use `pass` for empty body.

File: Lib/test/test_inspect.py [hunk 3]
```diff
@@ -1077,6 +1077,51 @@ class C(metaclass=M):
...
+        def function(): ''
+        a_lambda = lambda: ''
+        self.assertFalse(inspect.isdatadescriptor(type(Test().instance_method)),
+                         'a instance method is not a data descriptor')
+        self.assertFalse(inspect.isdatadescriptor(classmethod),
```
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: Why you test `classmethod`? Note that the type of class method is not `classmethod`.

File: Lib/test/test_inspect.py [hunk 4]
```diff
@@ -1077,6 +1077,51 @@ class C(metaclass=M):
...
+                         'a class method is not a data descriptor')
+        self.assertFalse(inspect.isdatadescriptor(staticmethod),
```
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: Ditto.

File: Lib/test/test_inspect.py [hunk 5]
```diff
@@ -1077,6 +1077,51 @@ class C(metaclass=M):
         attrs = [a[0] for a in inspect.getmembers(C)]
         self.assertNotIn('missing', attrs)
 
+class TestIsDataDescriptor(unittest.TestCase):
```
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py:1080]: These tests are not ran. Add the test class in `run_unittest()` list below.

File: Lib/test/test_inspect.py [hunk 6]
```diff
@@ -1077,6 +1077,51 @@ class C(metaclass=M):
...
+    def test_custom_descriptors(self):
+        class NonDataDescriptor(object):
+            def __get__(self, value, type=None): ""
+        class DataDescriptor0(object):
+            def __set__(self, name, value): ""
+        class DataDescriptor1(object):
+            def __delete__(self, name): ""
+        class DataDescriptor2(object):
+            __set__ = None
+        self.assertFalse(inspect.isdatadescriptor(NonDataDescriptor),
```
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: This returns `False` because `NonDataDescriptor` is a class.

File: Lib/test/test_inspect.py [hunk 7]
```diff
@@ -1077,6 +1077,51 @@ class C(metaclass=M):
...
+                         'class with only __get__ not a data descriptor')
+        self.assertTrue(inspect.isdatadescriptor(DataDescriptor0),
```
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: This test would fail if `TestIsDataDescriptor` run.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Lib/test/test_inspect.py (hunk 5 above)
Comment (aaronchall) [reply Lib/test/test_inspect.py:1080]: Added

**Comment by aaronchall** [comment]
I thought it was, "I didn't expect a kind of Spanish Inquisition!" not, I didn't expect the Spanish Inquisition!
(Nevertheless, I appreciate rigor in modifications to the best language of all time!)
Also, I have no idea how to resolve bedevere/news or appveyor's issues...

**Comment by aaronchall** [comment]
@serhiy-storchaka Do you want an entry in the Misc/NEWS.d directory? If I do that, should I remove the entry in Misc/NEWS?

**Comment by serhiy-storchaka** [comment]
`Misc/NEWS` no longer edited directly. You should add new entries into the `Misc/NEWS.d` directory  as individual files. See https://docs.python.org/devguide/committing.html#what-s-new-and-news-entries.

**Review by serhiy-storchaka — APPROVED** [review]
Thank you @aaronchall! LGTM. But since this is related to delicate matters, it would be nice if yet one core dev will take a look.

**Comment by benjaminp** [comment]
In general, this seems fine. My one nit is that rather than "and/or", let's just say "or". "and/or" is jargony.

**Comment by aaronchall** [comment]
@benjaminp done, however, the "appveyor" build failed - not sure why, appears to be irrelevant.

**Pull request merged**